### PR TITLE
Support custom chat endpoint without endpoint type set as llm judge

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/13538?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13538/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13538
```

</p>
</details>

### What changes are proposed in this pull request?

Support using a custom serving endpoint as a LLM judge for GenAI metrics like `answer_similarity`. Today, MLflow only supports serving endpoint that has endpoint type or task defined to be `llm/v1/chat` or `llm/v1/completion`. This works for FMAPI or some flavors, but not for model that does not have task concept e.g. LangChain.

This PR addresses this issue by defaulting to a chat task if the 

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Test data: 
```
import pandas as pd

eval_data = pd.DataFrame({
        "inputs": ["What is MLflow?"],
        "ground_truth": ["MLflow is an OSS platform for managing ML lifecycle."],
})
```

**Case 1. Custom Chat Endpoint**:

(The serving service is down so I cannot test it right now, instead, I validated the change by mocking `get_endpoint` API of the deploy client)

```
# Custom chat model (mocking with FMAPI until model serving service recovers)
from unittest import mock
from mlflow.deployments import get_deploy_client

real_client = get_deploy_client()
mock_client = mock.MagicMock(spec=real_client.__class__)
mock_client.get_endpoint.return_value = {}
mock_client.predict = real_client.predict

with mock.patch("mlflow.deployments.get_deploy_client", return_value=mock_client) as mock_get_deploy_client:
    mock_get_deploy_client.return_value = mock_client

    # Validate the mocking works
    from mlflow.deployments import get_deploy_client
    from mlflow.metrics.genai.model_utils import get_endpoint_type
    client = get_deploy_client()
    endpoint_details = client.get_endpoint("endpoints:/databricks-meta-llama-3-1-70b-instruct")
    assert "task" not in endpoint_details
    assert "endpoint_type" not in endpoint_details
    assert get_endpoint_type("endpoints:/databricks-meta-llama-3-1-70b-instruct") is None

    # Run evaluation
    with mlflow.start_run():
        results = mlflow.evaluate(
            "runs:/f71fc47873394a35843724165b772644/model",
            eval_data,
            targets="ground_truth",
            extra_metrics=[
                answer_similarity(model="endpoints:/databricks-meta-llama-3-1-70b-instruct"),
                answer_correctness(model="endpoints:/databricks-mixtral-8x7b-instruct")
            ]
        )
        print(f"Aggregated evaluation results: \n{results.metrics}")
```

> Aggregated evaluation results: 
{'answer_similarity/v1/mean': 4.0, 'answer_similarity/v1/variance': 0.0, 'answer_similarity/v1/p90': 4.0, 'answer_correctness/v1/mean': 5.0, 'answer_correctness/v1/variance': 0.0, 'answer_correctness/v1/p90': 5.0}

This should properly emulate the requirement, as the same code fails with MLflow 2.17.0:

> Aggregated evaluation results: 
{'answer_similarity/v1/mean': nan, 'answer_similarity/v1/variance': nan, 'answer_correctness/v1/mean': nan, 'answer_correctness/v1/variance': nan}

**Case 2. OpenAI**:
```
with mlflow.start_run():
    results = mlflow.evaluate(
        "runs:/f71fc47873394a35843724165b772644/model",
        eval_data,
        targets="ground_truth",
        extra_metrics=[
            answer_similarity(), # Default
            answer_correctness(model="openai:/gpt-4o-mini"),
        ]
    )
    print(f"Aggregated evaluation results: \n{results.metrics}")
```
> Aggregated evaluation results: 
{'answer_similarity/v1/mean': 4.0, 'answer_similarity/v1/variance': 0.0, 'answer_similarity/v1/p90': 4.0, 'answer_correctness/v1/mean': 5.0, 'answer_correctness/v1/variance': 0.0, 'answer_correctness/v1/p90': 5.0}

**Case 3: FMAPI**:
```
with mlflow.start_run():
    results = mlflow.evaluate(
        "runs:/f71fc47873394a35843724165b772644/model",
        eval_data,
        targets="ground_truth",
        extra_metrics=[
            answer_similarity(model="endpoints:/databricks-meta-llama-3-1-70b-instruct"),
            answer_correctness(model="endpoints:/databricks-mixtral-8x7b-instruct")
        ]
    )
    print(f"Aggregated evaluation results: \n{results.metrics}")
```
> Aggregated evaluation results: 
{'answer_similarity/v1/mean': 4.0, 'answer_similarity/v1/variance': 0.0, 'answer_similarity/v1/p90': 4.0, 'answer_correctness/v1/mean': 5.0, 'answer_correctness/v1/variance': 0.0, 'answer_correctness/v1/p90': 5.0}

**Case 4: External Model (proxy)**
```
# External Model
with mlflow.start_run():
    results = mlflow.evaluate(
        "runs:/f71fc47873394a35843724165b772644/model",
        eval_data,
        targets="ground_truth",
        extra_metrics=[
            answer_similarity(model="endpoints:/yuki-watanabe-openai-proxy-endpoint"),
            answer_correctness(model="endpoints:/yuki-watanabe-openai-proxy-endpoint"),
        ]
    )
    print(f"Aggregated evaluation results: \n{results.metrics}")
```
> Aggregated evaluation results: 
{'answer_similarity/v1/mean': 4.0, 'answer_similarity/v1/variance': 0.0, 'answer_similarity/v1/p90': 4.0, 'answer_correctness/v1/mean': 5.0, 'answer_correctness/v1/variance': 0.0, 'answer_correctness/v1/p90': 5.0}

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Support custom chat endpoint without endpoint type set as llm judge

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/metrics/genai/genai_metric.py [hunk 1]
```diff
@@ -99,12 +99,16 @@ def _extract_score_and_justification(text):
...
-    eval_model,
-    parameters,
+    payload: str,
+    eval_model: str,
+    parameters: Optional[Dict[str, Any]],
 ):
     try:
-        raw_result = model_utils.score_model_on_payload(eval_model, payload, parameters)
+        # If the endpoint does not specify type, default to chat format
+        endpoint_type = model_utils.get_endpoint_type(eval_model) or "llm/v1/chat"
```
Comment (BenWilson2) [inline mlflow/metrics/genai/genai_metric.py:108]: Safe assumption

File: mlflow/metrics/genai/model_utils.py [hunk 2]
```diff
@@ -9,9 +10,32 @@
 _logger = logging.getLogger(__name__)
 
 
+def get_endpoint_type(endpoint_uri: str) -> Optional[str]:
```
Comment (B-Step62) [inline mlflow/metrics/genai/model_utils.py:13]: note: just extracted from the original `score_model_on_payload` function to handle new requirement

File: mlflow/metrics/genai/model_utils.py [hunk 3]
```diff
@@ -9,9 +10,32 @@
...
+    endpoint = client.get_endpoint(endpoint_uri)
+    # TODO: Standardize the return type of `get_endpoint` and remove this check
+    endpoint = endpoint.dict() if isinstance(endpoint, BaseModel) else endpoint
+    return endpoint.get("task", endpoint.get("endpoint_type"))
+
+
 # TODO: improve this name
-def score_model_on_payload(model_uri, payload, eval_parameters=None):
-    """Call the model identified by the given uri with the given payload."""
+def score_model_on_payload(model_uri, payload, eval_parameters=None, endpoint_type=None):
```
Comment (B-Step62) [inline mlflow/metrics/genai/model_utils.py:37]: note: I will file a follow-up to rename `payload` to `prompt` (and elsewhere in the caller side). The `payload` is over-loaded within genai metrics code base, it is sometimes string (`content`) while also refers to the actual chat payload. This is very confusing.

Comment (BenWilson2) [reply mlflow/metrics/genai/model_utils.py:37]: Agreed on this point. Far too generic of a term when reading through these APIs :)

File: mlflow/metrics/genai/model_utils.py [hunk 4]
```diff
@@ -215,6 +218,32 @@ def _call_gateway_api(gateway_uri, payload, eval_parameters):
         )
 
 
+def _construct_payload_from_str(prompt: str, endpoint_type: str) -> Dict[str, Any]:
```
Comment (BenWilson2) [inline mlflow/metrics/genai/model_utils.py:221]: Thank you for this refactoring. So much cleaner.

**Review by BenWilson2 — APPROVED** [review]
Overall the approach LGTM and the refactoring really reduces the cognitive load on trying to follow dense conditional chains. Great stuff!

**Comment by B-Step62** [comment]
`mlflow-recipes` failure is not relevant to this PR: https://github.com/mlflow/mlflow/pull/13539

File: mlflow/metrics/genai/model_utils.py [hunk 5]
```diff
@@ -215,6 +218,32 @@ def _call_gateway_api(gateway_uri, payload, eval_parameters):
...
+    if endpoint_type == "llm/v1/completions":
+        return {"prompt": prompt}
+    elif endpoint_type == "llm/v1/chat":
+        return {"messages": [{"role": "user", "content": prompt}]}
+    else:
+        raise MlflowException(
+            f"Unsupported endpoint type: {endpoint_type}. If string input is provided, "
+            "the endpoint type must be 'llm/v1/completions' or 'llm/v1/chat'.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
```
Comment (harupy) [inline mlflow/metrics/genai/model_utils.py:235]: We don't need to worry about `embeddings`?

Comment (harupy) [reply mlflow/metrics/genai/model_utils.py:235]: I see, `_call_deployments_api` doesn't support embeddings.

File: mlflow/metrics/genai/model_utils.py [hunk 6]
```diff
@@ -215,6 +218,32 @@ def _call_gateway_api(gateway_uri, payload, eval_parameters):
...
+
+
+def _parse_response(response: Dict[str, Any], endpoint_type: Optional[str]) -> str:
```
Comment (harupy) [inline mlflow/metrics/genai/model_utils.py:238]:
```suggestion
def _parse_response(response: Dict[str, Any], endpoint_type: Optional[str]) -> Union[Optional[str], Dict[str, Any]:
```

nit

**Review by harupy — APPROVED** [review]
A few nits for type hints, otherwise LGTM!

File: mlflow/metrics/genai/model_utils.py (hunk 6 above)
Comment (harupy) [reply mlflow/metrics/genai/model_utils.py:238]: Not a blocker, can be fixed later.

File: mlflow/metrics/genai/model_utils.py (hunk 5 above)
Comment (B-Step62) [reply mlflow/metrics/genai/model_utils.py:235]: Yup, this is dedicated for llm-as-a-judge metric now so we don't need embedding🙂

**Comment by daniellok-db** [comment]
thanks for the change! this makes a lot of sense 😄

**Pull request merged**